lib/protocol/websocket/close_frame.rb

```diff
@@ -29,11 +29,41 @@ class CloseFrame < Frame
 			def unpack
 				data = super
 				
-				return data.unpack(FORMAT)
+				case data.length
+				when 0
+					[nil, ""]
+				when 1
+					raise ProtocolError, "invalid close frame length!"
+				else
+					code, reason = *data.unpack(FORMAT)
+					
+					case code
+					when 0 .. 999, 1005 .. 1006, 1015, 5000 .. 0xFFFF
+						raise ProtocolError, "invalid close code!"
+					when 1004, 1016 .. 2999
+						raise ProtocolError, "reserved close code!"
+					end
+					
+					reason.force_encoding(Encoding::UTF_8)
+					
+					unless reason.valid_encoding?
+						raise ProtocolError, "invalid UTF-8 in close reason!"
+					end
+					
+					[code, reason]
+				end
 			end
 			
 			def pack(code, reason)
-				super [code, reason].pack(FORMAT)
+				if code
+					unless reason.encoding == Encoding::UTF_8
+						reason = reason.encode(Encoding::UTF_8)
+					end
+					
+					super [code, reason].pack(FORMAT)
+				else
+					super "".b
```

Comment:
Is it valid to have a close frame with no code or reason?

Reply:
Yep, it is optional. https://datatracker.ietf.org/doc/html/rfc6455#section-5.5.1

> The Close frame MAY contain a body

> If there is a body, the first two bytes of the body MUST be a 2-byte unsigned integer